### PR TITLE
feat: dedup & cached build performance improvements

### PR DESCRIPTION
**Pull request by mndeveci**
#### Which issue(s) does this change fix?
#3093
#2688

#### Why is this change necessary?
Previously, SAM CLI improved build performance by introducing de-duplicated & cached builds. These improvements are using copying previously built resources to skip some of the build operations. But runtimes like `nodejs` might contain lots of files inside `node_modules` folder which makes copy operation slow.

#### How does it address the issue?
In order to address this issue, this PR has following improvements;
- For de-duplicated builds (where multiple functions uses same `CodeUri`) we will be building into a shared folder and point all of the functions to that folder rather than copying first built function into rest of them.
- For cached builds, we will be symlinking previously built artifacts to the target folder, instead of copying them.
- For Auto Dependency Layer (SAM Accelerate feature, where dependencies of a function is moved to a separate layer), we will be symlinking the dependencies to the target layer folder.

#### What side effects does this change have?
Certain operating systems might not support symlinking correctly. Because of that, we are checking if any errors are thrown with symlink operation, if so we will fall-back to copy. 
This feature will also be released behind a feature flag to address some issues before marking it as GA.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/build/build_graph.py [hunk 1]
```diff
@@ -625,7 +626,10 @@ def get_build_dir(self, artifact_root_dir: str) -> str:
         Return the directory path relative to root build directory
         """
         self._validate_functions()
-        return self.functions[0].get_build_dir(artifact_root_dir)
+        build_dir = self.functions[0].get_build_dir(artifact_root_dir)
+        if is_experimental_enabled(ExperimentalFlag.BuildImprovements22) and len(self.functions) > 1:
+            build_dir = f"{build_dir}-Shared"
```
Comment (sriram-mv) [inline samcli/lib/build/build_graph.py:631]: How is this postfix "Shared" determined?

Comment (mildaniel) [reply samcli/lib/build/build_graph.py:631]: When determining the artifacts directory that we will have our resultant build in, before we would copy all the contents in the same code URI (technically the same build definition) over and over. Now if there are multiple functions with the same build definition we will just put them into one single shared artifacts directory.

**Comment by mildaniel** [comment]
> I wonder if we can recommend this option/flag based on the runtime...if the project is using node, we can probably add a hint to use this in the CLI.

These changes aren't necessarily runtime-specific except for the accelerate related changes which are only applicable to Python and esbuild.

File: samcli/lib/build/build_graph.py (hunk 1 above)
Comment (sriram-mv) [reply samcli/lib/build/build_graph.py:631]: I see, putting up that comment above the line and also making the suffix a constant string thats defined and importable would be good.

**Review by sriram-mv — APPROVED** [review]

File: samcli/lib/build/build_graph.py [hunk 2]
```diff
@@ -625,7 +627,12 @@ def get_build_dir(self, artifact_root_dir: str) -> str:
         Return the directory path relative to root build directory
         """
         self._validate_functions()
-        return self.functions[0].get_build_dir(artifact_root_dir)
+        build_dir = self.functions[0].get_build_dir(artifact_root_dir)
+        if is_experimental_enabled(ExperimentalFlag.BuildPerformance) and len(self.functions) > 1:
```
Comment (mildaniel) [inline samcli/lib/build/build_graph.py:631]: Add message to customers about this change and add a link to a doc about all the changes that this includes.

**Review by qingchm — APPROVED** [review]
Discussion for review session has been captured in the comments above. The main focus would be to release this feature in the next release and then construct a corresponding documentation link explaining on the experimental features. With that customers can better understand the improvements.

**Pull request merged**